log4j-layout-template-json/src/main/java/org/apache/logging/log4j/layout/template/json/util/JsonWriter.java

```diff
@@ -513,18 +509,15 @@ public void writeString(final StringBuilderFormattable formattable) {
             writeNull();
         } else {
             stringBuilder.append('"');
-            formattableBuffer.setLength(0);
-            formattable.formatTo(formattableBuffer);
-            final int length = formattableBuffer.length();
-            // Handle max. string length complying input.
-            if (length <= maxStringLength) {
-                quoteString(formattableBuffer, 0, length);
-            }
+            int position = stringBuilder.length();
+            formattable.formatTo(stringBuilder);
+            final int length = stringBuilder.length() - position;
             // Handle max. string length violating input.
-            else {
-                quoteString(formattableBuffer, 0, maxStringLength);
+            if (length > maxStringLength) {
+                stringBuilder.setLength(position + maxStringLength);
                 stringBuilder.append(quotedTruncatedStringSuffix);
             }
+            StringBuilders.escapeJson(stringBuilder, position);
             stringBuilder.append('"');
```

Comment:
I haven't run all the tests yet, I may have broken something

Reply:
My comment above regarding `StringBuilders.escapeJson()` applies here too.